File: Week 2/id_445/144.二叉树的前序遍历.cpp
/*
 * @lc app=leetcode.cn id=144 lang=cpp
 *
 * [144] 二叉树的前序遍历
 */
#include <stack>
#include <vector>
using namespace std;
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
// solution 1 recursively
    // vector<int> preorderTraversal(TreeNode* root) {
    //     vector<int> res;
    //     recur(root,res);
    //     return res;
    // }
    // void recur( TreeNode* node,vector<int> &res){
    //     if (node==NULL)
    //     {
    //         return;
    //     }
    //     res.push_back(node->val);
    //     recur(node->left,res);
    //     recur(node->right,res);
        
    // }
    // solution2 iteratively
     vector<int> preorderTraversal(TreeNode* root) {
         vector<int> res;
         stack<TreeNode> s;
         while (true)
         {
             while (root)
             {
                 res.push_back(root->val);
                 s.push(*root);
                 root=root->left;

             }
             if(s.empty()) return res;
             root=s.top().right;
             s.pop();
             
         }
         
     }
};
// @lc code=end

File: Week 2/id_445/144.二叉树的前序遍历.py
#
# @lc app=leetcode.cn id=144 lang=python3
#
# [144] 二叉树的前序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # solution 1 recursively
    # def preorderTraversal(self, root: TreeNode) -> List[int]:
    #     res=[]
    #     self.recur(root,res)
    #     return res
    # def recur(self,node,res):
    #     if not node:
    #         return
    #     res.append(node.val)
    #     self.recur(node.left,res)
    #     self.recur(node.right,res)
    # solution 2 iteratively
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res,stack=[],[]
        while True:
            while  root:
                res.append(root.val)
                stack.append(root)
                root=root.left
            if not stack:
                return res
            root=stack.pop().right
            

        
# @lc code=end

